# EDA — Customer Churn (dataset histórico)

Exploración inicial del dataset `customer_churn_historical.csv` (7043 clientes, 21 columnas).
Objetivo: identificar problemas de calidad de datos y definir las decisiones
de preprocesamiento que debe implementar el pipeline.

Análisis por segmentos (churn por contrato, tipo de internet, método de pago):
pendiente, ver `02_eda_segmentos.ipynb`.

In [16]:
import pandas as pd

df = pd.read_csv("../data/raw/customer_churn_historical.csv")

In [17]:
df.shape

(7043, 21)

In [18]:
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,6684-LWVH,Male,0,No,No,4,Yes,Yes,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,One year,No,Mailed check,27.56,107.47,No
1,7027-JIFO,Male,0,No,No,35,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Month-to-month,Yes,Mailed check,30.52,1020.47,No
2,5981-VOQO,Female,0,No,No,45,Yes,Yes,DSL,No,...,No,No,Yes,No,Month-to-month,Yes,Bank transfer (automatic),66.87,2686.35,No
3,7266-HYDM,Male,0,Yes,No,31,Yes,Yes,Fiber optic,No,...,No,No,No,Yes,Month-to-month,Yes,Bank transfer (automatic),86.95,2536.96,Yes
4,2821-JDLS,Female,0,Yes,No,5,Yes,Yes,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,30.18,139.12,No


In [19]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   str    
 1   gender            7043 non-null   str    
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   str    
 4   Dependents        7043 non-null   str    
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   str    
 7   MultipleLines     7043 non-null   str    
 8   InternetService   7043 non-null   str    
 9   OnlineSecurity    7043 non-null   str    
 10  OnlineBackup      7043 non-null   str    
 11  DeviceProtection  7043 non-null   str    
 12  TechSupport       7043 non-null   str    
 13  StreamingTV       7043 non-null   str    
 14  StreamingMovies   7043 non-null   str    
 15  Contract          7043 non-null   str    
 16  PaperlessBilling  7043 non-null   str    
 17  Paymen

In [20]:
df[df["TotalCharges"].isna()]

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
95,7731-MHBI,Female,0,Yes,No,0,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Month-to-month,Yes,Electronic check,24.03,NaN,No
781,8834-VPIU,Male,1,Yes,Yes,0,Yes,Yes,Fiber optic,No,...,Yes,No,No,No,Month-to-month,Yes,Bank transfer (automatic),83.01,NaN,No
854,3651-KMCX,Female,0,No,Yes,0,Yes,Yes,DSL,No,...,No,Yes,No,Yes,Month-to-month,No,Credit card (automatic),68.81,NaN,No
867,9404-ZJPR,Female,0,Yes,Yes,0,Yes,Yes,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Credit card (automatic),75.01,NaN,No
882,3039-OWAI,Male,1,No,No,0,Yes,No,DSL,No,...,Yes,Yes,Yes,Yes,Month-to-month,Yes,Credit card (automatic),74.50,NaN,Yes
1013,6432-UDAD,Male,0,Yes,Yes,0,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Month-to-month,Yes,Credit card (automatic),24.19,NaN,Yes
1172,9097-BQWV,Female,0,Yes,No,0,No,No phone service,Fiber optic,Yes,...,Yes,Yes,No,Yes,Two year,Yes,Credit card (automatic),89.15,NaN,Yes
1842,8166-JZVI,Male,0,Yes,No,0,Yes,Yes,Fiber optic,Yes,...,No,Yes,No,Yes,Two year,No,Credit card (automatic),89.85,NaN,No
2112,2033-KDTO,Female,0,No,No,0,Yes,No,DSL,No,...,Yes,No,No,No,Month-to-month,Yes,Electronic check,59.50,NaN,Yes
2346,1069-TJKG,Male,1,Yes,No,0,Yes,No,Fiber optic,No,...,Yes,No,No,Yes,One year,No,Electronic check,84.69,NaN,Yes


In [21]:
(df["tenure"] == 0).sum()

np.int64(30)

In [22]:
df[(df["tenure"] == 0) & (df["TotalCharges"].notna())]

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
135,1326-PAXN,Female,0,No,No,0,Yes,No,Fiber optic,No,...,Yes,Yes,No,No,Month-to-month,Yes,Bank transfer (automatic),81.13,0.0,Yes
1887,9354-MLMH,Female,0,No,No,0,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Credit card (automatic),67.56,0.0,No
4790,5341-AISN,Female,0,No,Yes,0,Yes,No,Fiber optic,No,...,Yes,Yes,Yes,Yes,Month-to-month,No,Credit card (automatic),102.11,0.0,No
5751,4532-FKSI,Male,0,No,Yes,0,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,One year,Yes,Bank transfer (automatic),27.46,0.0,No


## Hallazgo 1: valores faltantes en TotalCharges

26 valores nulos, todos en clientes con tenure = 0.
Hay 30 clientes con tenure = 0: 26 tienen TotalCharges nulo y 4 tienen 0.0 explícito.
El nulo no es un error de carga, representa "todavía no facturó".

**Decisión para el pipeline:** imputar TotalCharges = 0 cuando falta.
No usar media ni mediana: asignaría ~1400 de historial de facturación
a clientes que recién ingresaron, justo el segmento con mayor churn.

In [23]:
df["InternetService"].value_counts()

InternetService
Fiber optic    3165
DSL            2329
No             1549
Name: count, dtype: int64

In [24]:
df["OnlineSecurity"].value_counts()

OnlineSecurity
No                     3203
Yes                    2291
No internet service    1549
Name: count, dtype: int64

In [25]:
for col in ["OnlineBackup", "DeviceProtection", "TechSupport", "StreamingTV", "StreamingMovies"]:
    print(col, (df[col] == "No internet service").sum())

OnlineBackup 1549
DeviceProtection 1549
TechSupport 1549
StreamingTV 1549
StreamingMovies 1549


## Hallazgo 2: categorías redundantes en servicios de internet

1549 clientes (22%) tienen InternetService = "No".
Ese mismo grupo aparece como "No internet service" en 6 columnas:
OnlineSecurity, OnlineBackup, DeviceProtection, TechSupport,
StreamingTV, StreamingMovies.

Al aplicar OneHotEncoder se generarían 6 variables binarias idénticas
entre sí e idénticas a InternetService_No: colinealidad perfecta,
coeficientes de la regresión logística no interpretables.

**Decisión para el pipeline:** recodificar "No internet service" como "No"
en las 6 columnas. InternetService queda como única variable que informa
si el cliente tiene internet.

In [26]:
df["Churn"].value_counts(normalize=True)

Churn
No     0.736334
Yes    0.263666
Name: proportion, dtype: float64

## Hallazgo 3: desbalance de clases

Churn = No: 73.6%

Churn = Yes: 26.4%

Un clasificador trivial que prediga siempre "No" alcanza 73.6% de accuracy
sin aprender nada. La accuracy no discrimina un modelo útil de uno inútil
en este dataset.

**Decisión para el pipeline:** métrica principal recall sobre la clase
positiva, acompañada de precision, F1 y ROC-AUC. Evaluar class_weight='balanced'
en los modelos y ajustar el umbral de decisión.

**Justificación de negocio:** el falso negativo (no detectar a quien se va)
cuesta el cliente completo más el costo de readquisición. El falso positivo
(retener a quien se quedaba) cuesta solo el margen del descuento.
La asimetría favorece recall sobre precision.